### ISSUE TITLE
Status 400 and reason BadDeviceToken

### ISSUE THREAD

**Issue by halvini** [parse-community/parse-server-push-adapter#87]
I've updated the parse-server to the latest 2.5.3 version and parse-server-push-adapter v 2.0.0 and now my error logs are full of these messages:
```
3|parse-wr | node-pre-gyp
3|parse-wr |  
3|parse-wr | ERR!
3|parse-wr |  
3|parse-wr | parse-server-push-adapter APNS
3|parse-wr |  APNS error transmitting to device 28c89bdc6a0e7a6454fb0cc446229a0841eecdea955533d3c549a658bb15fc98 with status 400 and reason BadDeviceToken
```

There are so many of them that I can't see anything else.

The questions are:
1) How can I ignore these errors in my pm2 config file?
2) Is there a way to remove bad device tokens automatically from Installation Collection?

**Comment by flovilmart** [comment]
> How can I ignore these errors in my pm2 config file?

There is no particular way to ignore those, they are produced when running with VERBOSE

> Is there a way to remove bad device tokens automatically from Installation Collection?

For now there isn't

**Comment by halvini** [comment]
Hey @flovilmart, my VERBOSE setting is disabled, but my logs are filled with these errors by 90%.

How do I turn them off?

**Comment by flovilmart** [comment]
After looking into the repo, the error is reported by the logger as an error log, you won’t be able to disable it at that time.

**Comment by miguel-s** [comment]
Hi @flovilmart, we are running our parse server with `logLevel: 'none'`, but we are still getting these errors logged since we upgraded to 2.5.3, is this intended?

**Comment by halvini** [comment]
Looks like they just forgot to set the right setting for these errors.

I've downgraded to an earlier version.

**Comment by halvini** [comment]
Any news on this issue? Is there a way to switch the Push errors off? 

Or is there a way to automatically erase all Installations with Bad Device Tokens?

**Comment by funkenstrahlen** [comment]
As far as I know in the latest version there is a way to automatically delete bad tokens.

**Comment by halvini** [comment]
Do you know any adequate way to find the bad device tokens and erase them?

We have a bunch of them, and they only increase in numbers, especially the ones for Androids. They never self-delete even if the app is reinstalled.

**Comment by wicked-wei** [comment]
Hi @funkenstrahlen can you please tell the way to automatically delete bad tokens? thanks a lot.

**Comment by halvini** [comment]
@wicked-wei , our tests for this PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS option showed that:
1) even the VALID iOS deviceTokens are erased for some reason
2) the old, invalid, unused android deviceTokens are not erased at all

**Comment by wicked-wei** [comment]
@halvini thanks for your comment. I haven't seen the problem you mentioned yet but I will keep eyes on it.
And, the code that PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS option works is at  parse-server/src/StatusHandler.js, you can check it out for more detail.

**Comment by flovilmart** [comment]
@halvini can you please point out what part of the logic there is not working as expected?

**Comment by halvini** [comment]
@flovilmart , the issue is described in this ticket:
https://github.com/parse-community/Parse-SDK-iOS-OSX/issues/1224

We are 100% sure that PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS flag cleans up even VALID ios installations. Why? I don't know. I've checked the StatusHandler, I've checked the APNS.js. I don't see anything unusual out there, but I believe that when you send many many push notifications to different installations, probably APNS returns BadDeviceToken error even for valid installations. It doesn't happen immediately. For instance, my own valid installation was unset by this flag in a week after I turned it on. So, probably, if there is a timeout, or any other connection or delivery error, the APNS thinks that device is either Unregistered, or Bad.

**Comment by wicked-wei** [comment]
@halvini you can also check that if there is an unexpected mixed using of both apple sandbox and production push on a same environment(or database). This may cause unexpected deletion of deviceToken.

**Comment by halvini** [comment]
@wicked-wei the sandbox was used by 20 beta-testers only. 

When we set PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS flag, nothing bad happened at first, but then in 3 days we started getting complaints from both sandbox users and production users. Why were the production users affected by this flag?

**Comment by wicked-wei** [comment]
@halvini  because sandbox and production use different deviceToken. If any chance you send a production deviceToken to a sandbox apns server, it will report that token is BadToken. Reverse is the same.

**Comment by halvini** [comment]
@wicked-wei ok, so why do production users lost their deviceToken data, if we send push notifications from cloud code only?

**Comment by wicked-wei** [comment]
@halvini the config looks ok, but I am not sure if parse server can tell a installation deviceToken is production or sandbox. For the js documents does not mention that, maybe it will send a token to both production or sandbox? I am not sure because I never do that.

**Comment by halvini** [comment]
@wicked-wei , this is all quite informative, but it doesn't resolve the issue of undefined deviceTokens for valid ios installations. 

Your idea is to change the development certificate to a new bundle id, am I right? How can I be sure that that's the reason? I can't risk my production environment again. I've lost too much data when I set the PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS flag. It worked for a few days only but resulted in a huge mess.

**Comment by halvini** [comment]
Did anyone find a solution to this issue?

I'd like to set the PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS to 1 without losing valid iOS installations.

**Comment by flovilmart** [comment]
@alexblack  @jeffreyjackson the only way to know if it's safe to use is for you guys to use it right&

**Comment by alexblack** [comment]
Well, there are some findings above, sounds risky, we'll find another way.

**Comment by alexblack** [comment]
Fair point.  I should share my other feedback - also not too excited about the feature in general, if device tokens are erased then it seems to make it harder to distinguish installations with no device token vs ones that were erased.

This combined with the lack of real errors/transparency from parse push means we'll eventually just integrate with apple and google, then we'll have real errors back, and we can just mark installations as unreachable when we learn that from the source.

**Comment by flovilmart** [comment]
> also not too excited about the feature in general, if device tokens are erased then it seems to make it harder to distinguish installations with no device token vs ones that were erased.

This is what you were expecting originally, that this feature is working correctly, not sure where you're coming from with that.

> This combined with the lack of real errors/transparency from parse push means we'll eventually just integrate with apple and google, then we'll have real errors back, and we can just mark installations as unreachable when we learn that from the source.

I'm open for improving the project / provide a better feature but without community feedback this is unfortunately impossible. 

Opensource doesn't mean we the maintainers have to take all the risks and responsibility.

The community pressed to have a cleanup invalid token implemented, IIRC you were part of it, I spent a shit load of hours on it, a tentative implementation has been available for a while and now you're pressing again today to have something else implemented. Do you understand how unfair, unethical and unwelcome those commentaries are?

**Comment by alexblack** [comment]
Sorry. Just sharing my feedback. :(

**Comment by alexblack** [comment]
I don't remember pressing to have device tokens erased, but, if I did, and now I realized that having them not erased would be better, I apologize for my lack of foresight.

**Comment by flovilmart** [comment]
> Any update here? is PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS safe to use? Are there docs on it?

@alexblack that's not what I call sharing feedback, but waiting-for-someone-else-to-do-some-work-for-me.

Sharing feedback would have been to open issues, discuss potential improvements and most importantly, identify if there's any issue with the approach. What would be the best solution for you?

Now, we gather all error reports from the adapter, and list the installations that fail. Removing their token is what's expected from GCM or APNs. What's missing?

**Comment by alexblack** [comment]
I put in some hours trying to make a PR in the past.  I failed.  That seems like more than `waiting-for-someone-else-to-do-some-work-for-me.`

**Comment by flovilmart** [comment]
If you have trouble opening a PR, we, the maintainers are here to help

**Comment by alexblack** [comment]
> Now, we gather all error reports from the adapter, and list the installations that fail. Removing their token is what's expected from GCM or APNs. What's missing?

At the risk of being unethical and unfair, while this is a huge step forward (nice work), what ultimately I am looking for is:

1. for installations to be marked unreachable when they are unreachable, then we can avoid sending pushes to them (and to be able to distinguish different cases like installations that *never* had a device token etc)
2. to be able to see for any (push, installation) pair what the result was from the source (GCM, APN). eg did it succeed, did it fail, whats the error.

**Comment by alexblack** [comment]
1. seems like it'd be easy for (me, or someone else) to add. eg set a boolean installation.unreachable=true instead of erasing the device token.

2. I think what would be useful here is to have a new collection, `push_installation`, with columns `push_id`, `installation_id`, `result`

My pr was moving in that direction, and the sense I got from you was that this was not a direction you wanted to go.

Given that the errors are logged, I guess if they include the push id, and the installation id, then I could write something in our server to collect them, and put them into our own table, but it doesn't seem like a great solution.

**Comment by flovilmart** [comment]
> My pr was moving in that direction, and the sense I got from you was that this was not a direction you wanted to go.

Yes we need to collect the errors in a _PushStatus object and not in another collection. Our role also as maintainers is to maintain YOUR code once we put it in, so we have to weigh in and our on those.

> seems like it'd be easy for (me, or someone else) to add. eg set a boolean installation.unreachable=true instead of erasing the device token.

What would be the benefit over removing the deviceToken? Imagine the device token gets updated, what do we do with the unreachable? How do we handle conflicts if s/o already uses 'unreachable'? What if the installation is reachable but wrongly flagged?

**Comment by alexblack** [comment]
> What would be the benefit over removing the deviceToken? 

To distinguish these installations from ones which never had a deviceToken.  Some of our installations do not have deviceTokens, we need to make sure that doesn't happen often, and that we have a way to fix it for our users in that situation.

**Comment by alexblack** [comment]
> Yes we need to collect the errors in a _PushStatus object and not in another collection.

Makes sense.  If it was a map in there of `installationId` -> `result` that would be very useful.

**Comment by alexblack** [comment]
> Imagine the device token gets updated, what do we do with the unreachable?

It should be erased.

> How do we handle conflicts if s/o already uses 'unreachable'?

What is `s/o`?

> What if the installation is reachable but wrongly flagged?

Seems like this issue happens with the existing solution too.  Eg what if the installation is reachable, but you erased the device token.

At least with `unreachable` you could choose to re-attempt if you wanted.

**Comment by flovilmart** [comment]
>  I guess if they include the push id, and the installation id

It should be as is 
https://github.com/parse-community/parse-server/blob/57da2de/src/StatusHandler.js#L249

> Some of our installations do not have deviceTokens, we need to make sure that doesn't happen often, and that we have a way to fix it for our users in that situation.

Not sure what's to fix there, ultimately deleting the device tokens is just a way to ensure the delivery rates are close to 1.0 in your _PushStatus. Which in the end doesn't really matter.

> Seems like this issue happens with the existing solution too. Eg what if the installation is reachable, but you erased the device token.

I'm not saying the current solution is perfect, nor can't be improved. but in both cases, what you want is just tracking the errors in the _PushStatus object. THat you can use the _PushStatus object to update your installations to add an 'unreachable' flag periodically is your own prerogative. So just tracking the errors/responses into the _PushStatus would be probably OK.

**Comment by alexblack** [comment]
> Not sure what's to fix there, ultimately deleting the device tokens is just a way to ensure the delivery rates are close to 1.0 in your _PushStatus. Which in the end doesn't really matter.

If device tokens are erased, then we lose the ability to distinguish two different failure modes:

1. The device was reachable, its not not
2. The device failed to get a device token.

Distinguishing these two cases is important for us.

**Comment by flovilmart** [comment]
This is why ultimately because we have two different goals, we need to settle on a common ground which is simply to track the invalid deviceTokens. Which is done today through a log line, and could be put into the _PushStatus object.

**Comment by alexblack** [comment]
> So just tracking the errors/responses into the _PushStatus would be probably OK.

If errors/responses are added to _PushStatus in a map keyed by installationId, that'd be super helpful!

**Comment by alexblack** [comment]
> This is why ultimately because we have two different goals, we need to settle on a common ground 
which is simply to track the invalid deviceTokens.

I'm simply responding to your question `What would be the benefit over removing the deviceToken?`

If helping support my goal here doesn't fit in with the project's aims or priorities, fair enough.

**Comment by alexblack** [comment]
I tried before (along slightly different lines), I frankly sucked at it, so for now I'm not going to try again.  Sorry :(

**Comment by flovilmart** [comment]
That was really a waste of my time... thanks a lot

**Comment by alexblack** [comment]
I'd love for us to contribute somehow.  

1. Is there a way we could contribute financially on an ongoing basis, perhaps by paying for an existing (or additional) cloud service of some type?

2. I think we can get a PR done, this would be a good idea for one of the guys on my team to help out with, instead of us integrating directly with Google/Apple. 

Today a `_PushStatus` looks like this:

```
    {
      "objectId": "vT77c6QI3W",
      "pushTime": "2018-05-14T14:44:09.484Z",
      "query": "{\"user\":{\"$inQuery\":{\"where\":{\"account\":{\"__type\":\"Pointer\",\"className\":\"Account\",\"objectId\":\"Qq6tfNsUoR\"}},\"className\":\"_User\"}},\"objectId\":{\"$ne\":\"Y1jVkY4LZp\"},\"deviceType\":{\"$ne\":\"web\"},\"lastSeenAt\":{\"$gt\":{\"__type\":\"Date\",\"iso\":\"2018-04-14T14:44:09.444Z\"}}}",
      "payload": "{\"pushType\":4,\"pushVer\":1,\"fromInstallation\":\"Y1jVkY4LZp\",\"timestamp\":1526309053323,\"aps\":{\"content-available\":1},\"isId\":\"4299c240-5785-11e8-9be6-a7fdcaf3dcac\"}",
      "source": "rest",
      "status": "succeeded",
      "numSent": 0,
      "pushHash": "d41d8cd98f00b204e9800998ecf8427e",
      "createdAt": "2018-05-14T14:44:09.484Z",
      "updatedAt": "2018-05-14T14:44:09.770Z",
      "numFailed": 1,
      "failedPerType": {
        "android": 1
      },
      "ACL": {
        
      }
    }
```

Would the idea be to add a field to it like this?

```
    {
      "objectId": "vT77c6QI3W",
      "pushTime": "2018-05-14T14:44:09.484Z",
      "resultsByInstallationId": {
        "VvNFL12st4": { 
          succeeded: true, 
          result: {
            "message_id": "0:1526665330066787%b4783826f9fd7ecd"
          },
        "LBsWhDeAMn": { 
          succeeded: false, 
          result: {
            "error": "NotRegistered"
          }
       }        
   }
```

**Comment by flovilmart** [comment]
I don’t believe there’s the installationId back from the push adapter, only the deviceToken

**Comment by alexblack** [comment]
I'm pretty sure that a PR could attribute the result back to the installation id, and that seems like the right way to do it.  

If that was the case, does that new field `resultsByInstallationId` look like what we'd want? (The name might be a bit verbose :)

**Comment by flovilmart** [comment]
So go ahead open he PR. Add the tests, run it in your test environment. Ensure it fits what you need. We’ll discuss after.

**Comment by alexblack** [comment]
Could we not discuss the requirements up front? If that is not what you'd like to see, then best to figure that out now.

**Comment by flovilmart** [comment]
in all those discussions you would have had 10 times the time to implement this and run it, and see if it works for your use case. My gut feeling tells me that it will take you multiple iterations and days running in production to see if this works properly.

**Comment by alexblack** [comment]
For me, I think it'd make sense for us to have some tentative agreement on what the new field in `_PushStatus` looks like before getting started.

(Also, we can't get started right away regardless...)

**Comment by flovilmart** [comment]
We have an agreement that we want to track errors coming back from the push adapter into the _PushStatus object, so a user can at a later time identify which installation has an invalid deviceToken.

I believe this is clear enough as a description, and his gives you the freedom to implement it the way you feel is best. Also if uou’re Not able to start it, either immediately or in a foreseeable future, you don’t *really* *need* that feature, nor to discuss it this much is such short amount of time. Again I could be mistaken, but it feels nothing has much moved forward and we’ve both wasted our time on words rather than code, given that what you expect would pretty much take only a few minutes to write and about an hour of adding tests.

**Comment by alexblack** [comment]
I hear you.  

For me, if we had tentative agreement on what the field looked like in `_PushStatus` that would really help.   Without that, I worry that a PR would get rejected or require significant rework if it missed the mark on what this field should look like.

**Comment by flovilmart** [comment]
And so what if the PR is rejected or require rework? Does it never happen that at code review a better solution arises? Or that when implementing you realize things don’t fit the pre agreed mold? What if the solution is just to write a properly structured log because it works better for you? Wouldn’t it be a waste? Do you believe the time we spend here is not more wasteful than the time it would take to implement, and rework the said PR?

**Comment by alexblack** [comment]
I thought if we chatted briefly for a few minutes we'd land on what we wanted to see in `_PushStatus`.   Seemed worth it to me.  But I can appreciate if that doesn't seem worth it to you.

**Comment by flovilmart** [comment]
@halvini please update to the latest version of parse-server that now uses the HTTP/2 interface to APNS. It should yield significantly less logs and open a new issue if it persists.

@alexBlack, I stated my point loud and clear, and you should probably take what I say and work from it. I am not against your proposal, and waiting for an implementation as well as suffisant testing on your side that would ensure we’re not doing all of that for nothing.

**Comment by alexblack** [comment]
@flovilmart if we had some tentative agreement on what we'd like to see in `_PushStatus` it would give me the confidence I needed to proceed.  

If we can't get agreement there, I'd still hope we can contribute to this great project somehow.  Its definitely a huge help for us.

**Comment by flovilmart** [comment]
Please stop it. What are you expecting? That i’ll Sign off right now on your prospective PR because we agreed on the naming? Ok the name is approved! Now if you want to implement it go ahead!

**Comment by alexblack** [comment]
> What are you expecting?

```
"resultsByInstallationId": {
        "VvNFL12st4": { 
          succeeded: true, 
          result: {
            "message_id": "0:1526665330066787%b4783826f9fd7ecd"
          },
        "LBsWhDeAMn": { 
          succeeded: false, 
          result: {
            "error": "NotRegistered"
          }
       } 
```

I was hoping for some feedback that this looked like what you wanted, or, it didn't.

**Comment by flovilmart** [comment]
Go ahead with that, and run in on your infra for a week

**Comment by halvini** [comment]
@flovilmart thanks for the response! 

Guys, please don't shoot each other :)

**Comment by pcegarra** [comment]
same error here
`backend_1  | node-pre-gyp ERR! parse-server-push-adapter APNS APNS error transmitting to device 7bc3dc72efe2538299e8c775ds58dd0a9a8e8b50794d2 with status 410 and reason Unregistered
backend_1  | node-pre-gyp ERR! parse-server-push-adapter APNS APNS error transmitting to device 5b7c2580e1d0bd89d5399338b2c470c0c37c68b3862fc7 with status 400 and reason BadDeviceToken`

**Comment by flovilmart** [comment]
@pcegarra ensure you are using the latest parse server, and the device token is actually valid for your certificate pair.

**Comment by flovilmart** [comment]
Can you provide the logs when running with VERBOSE=1? Also the state of the associated _PushStatus object for this push?

**Comment by pcegarra** [comment]
Ok, im going to look with verbose=1, but i dont know how to see the state of the  associated _PushStatus object..
Can you explain it with other words please? thanks!

**Comment by flovilmart** [comment]
Make a query to the _PushStatus table and get the most recent object. Or with the verbose logs, you should have the objectId associated with it.

**Comment by pcegarra** [comment]
Parse verbose log:

`backend_1  | verbose: _PushStatus a47E2wHCoa: sending push to installations with 1 batches
backend_1  | verbose: Sending push to 2
backend_1  | verbose: Sending push to 1
backend_1  | verbose: Sending push to 1
backend_1  | verbose: Sending push to 2
backend_1  | verbose: Sending push to 1
backend_1  | verbose: _PushStatus pdEUFObZ0g: sent push! 0 success, 1 failures
backend_1  | verbose: _PushStatus pdEUFObZ0g: needs cleanup devicesToRemove=[APA91bHrIxaqZ2mVv33U89Hsk9mzPQk1d2s_rN2Fw59EhmSU9hAJqgHcw3M2u1MFc3s-IgiYyhmk0Cw-ZWNj6n9g5ljzZGmHrTwBipYLtnNgnRAswc_a7IcC1_7aj0GFRk9cGVod2Sq9]
backend_1  | verbose: _PushStatus ztbkBDkM1z: sent push! 0 success, 1 failures
backend_1  | verbose: _PushStatus ztbkBDkM1z: needs cleanup devicesToRemove=[dG6GAAOXF_c:APA91bG7vKuW58ni0p3uWnSCEyZn_ZxqRrmG6FSZE3SZrRr9vNnIRwBAff9tZLwvgXbaKwTjoo__qdSdV1xQeajzY3TvVqtsB-NpyV7rhTeccMqAcg8cyw5UJynw_Dw7YqcseBITKGkY]
backend_1  | verbose: _PushStatus 44XCEpCDpc: sent push! 1 success, 1 failures
backend_1  | verbose: _PushStatus 44XCEpCDpc: needs cleanup devicesToRemove=[cFcF8b-oBEw:APA91bF3aftS13tRQduyF1lUvCr9pcRIV4h1E0J_RrQOKpKOBB9JlxI70MLaH2-EMP0Oh6Oh5iw-swxzF0v80KzfGFH7cUe0X-8wyOaiV1Mrz0sy-izI3HYEQvDRGJ0RNfvdoN6eU1hW]
backend_1  | verbose: _PushStatus MHGZ2UPH2F: sent push! 2 success, 0 failures
backend_1  | verbose: _PushStatus MHGZ2UPH2F: needs cleanup devicesToRemove=[]
backend_1  | verbose: _PushStatus a47E2wHCoa: sent push! 1 success, 0 failures
backend_1  | verbose: _PushStatus a47E2wHCoa: needs cleanup devicesToRemove=[]
backend_1  | verbose: _PushStatus KKcNQReTQJ: sent push! 1 success, 1 failures
backend_1  | verbose: _PushStatus KKcNQReTQJ: needs cleanup devicesToRemove=[finju-C12RI:APA91bGtTbGBXfoDegjganBKoBf_Z8mNo_A-dO18JdCudyQTOKk7qKmtuxpRvbRYG5qa_6cw7IvOGP2AYL1NI6eVvywjDgaUgM9MDsd0wfdxY85zveZ--FkphO1vlcI2xeXg0-J78jTW]
backend_1  | node-pre-gyp ERR! parse-server-push-adapter APNS APNS error transmitting to device 7bc3dc72efe2538299e8c7755fdf1a20ca01073eea32b58dd0a9a8e8b50794d2 with status 410 and reason Unregistered
backend_1  | verbose: _PushStatus tGmYrDZiaY: sent push! 0 success, 1 failures
backend_1  | verbose: _PushStatus tGmYrDZiaY: needs cleanup devicesToRemove=[7bc3dc72efe2538299e8c7755fdf1a20ca01073eea32b58dd0a9a8e8b50794d2]
backend_1  | node-pre-gyp ERR! parse-server-push-adapter APNS APNS error transmitting to device dc5dd60549ab890a42f3afb1d47b99d766d2fd8b84ba67152a55b1afcc839b69 with status 410 and reason Unregistered
backend_1  | verbose: _PushStatus KiPDDNApW3: sent push! 2 success, 1 failures
backend_1  | verbose: _PushStatus KiPDDNApW3: needs cleanup devicesToRemove=[dc5dd60549ab890a42f3afb1d47b99d766d2fd8b84ba67152a55b1afcc839b69]`


Query id for KiPDDNApW3
`{
    "results": [
        {
            "objectId": "KiPDDNApW3",
            "pushTime": "2018-05-26T12:23:40.842Z",
            "query": "{\"restaurant\":{\"__type\":\"Pointer\",\"className\":\"Restaurant\",\"objectId\":\"6FgXTZOtI0\"}}",
            "payload": "{\"alert\":\"¿Que hay de comer en \\\"Tayo’s\\\"?\",\"badge\":1,\"sound\":\"default\"}",
            "source": "rest",
            "status": "succeeded",
            "numSent": 2,
            "pushHash": "ed5d0f1208f198fc8d337b2105a53b28",
            "createdAt": "2018-05-26T12:23:40.842Z",
            "updatedAt": "2018-05-26T12:23:43.183Z",
            "numFailed": 1,
            "sentPerType": {
                "ios": 2
            },
            "failedPerType": {
                "ios": 1
            },
            "ACL": {}
        }
    ]
}`

**Comment by flovilmart** [comment]
From what I can see, everything is fine for that push. In your logs we see at least 3 different pushs intertwined, but, the one with the push status look good, as the expected number of pushs is 3 and the total send / failed is 3 as well

**Comment by mtrezza** [comment]
So I've tested `PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS` with development certificates and it worked just fine.

It takes me about 5 remote push notifications within 5 minutes until APNS reports a failed push and the  bad device token gets removed. It worked several times, the good device tokens untouched, the bad ones removed.

**Parse server push certificate configuration:**
```
push: {
        ios: [
            {
                pfx: 'apns_dev.p12',
                topic: 'com.myDomain.myApp',
                production: false
            }
        ]
    },
```

**Environment:**
parse-server v2.7.4

Is @halvini is the only one experiencing problems? That case sounds more like some kind of mix-up of production and development certificates / device tokens. Note that my parse server configs only includes one APNS certificate per server.

Maybe it has something to do with how parse server handles an array of push certificates. The [docs](http://docs.parseplatform.org/parse-server/guide/#2-configure-parse-server) do not clarify for me what the purpose of multiple push certificates is and how parse server decides which one to use:
> If you would like to support both the dev and prod certificates, you can provide an array of configurations...

**Comment by mtrezza** [comment]
@flovilmart I've tested `PARSE_SERVER_CLEANUP_INVALID_INSTALLATIONS` with development certificates in APNS sandbox and it worked just fine.

It always takes me about 5 remote push notifications within 5 minutes until APNS reports a failed push and the bad device token gets removed. It worked several times, the good device tokens untouched, the bad ones removed.

**Parse server push configuration:**
```
push: {
        ios: [
            {
                pfx: 'apns_dev.p12',
                topic: 'com.myDomain.myApp',
                production: false
            }
        ]
    },
```
**Environment:**
parse-server v2.7.4

@halvini's case sounds like some kind of mix-up of production and development APNS certificates / device tokens. Note that my parse server configs only includes one APNS certificate per server.

Looking at the code I suspect the following cause:

Each push certificate creates a provider. The list of providers is sorted by production certificates before development certificates. When sending a push notification, the production provider is tried first, if it fails the development provider is used. Now if a push using an APNS production device token that is sent through the production provider fails for any reason, the push is sent through the development provider, which will identify it as a bad device token. This leads to the removal of the token in parse server because it has been sent to the wrong environment. So in @halvini's case the APNS production server could simply have been unreachable, 3 days after activating the clean-up flag, causing the removal of the good device tokens.

If that is the cause of the problem, a solution could be to add a `production: Bool` field to the `Installation` document to flag the device token as either production or development. The push sending logic then has to be changed to send production tokens only to production providers and development tokens only to development providers.

Another solution would be to simply remove the possibility of adding development certificates to parse server in a production environment, i.e. remove the `production` key from the parse push configuration. A development / testing / staging environment should probably run its own instance of parse server, sandboxed away from any production environment anyway.

**Comment by mtrezza** [comment]
@flovilmart, this issue is closed, but can you confirm or dismiss my analysis above?

**Comment by flovilmart** [comment]
Yeah it’s possible that the errors reported when using multiple certificates in production may be problematic. I’ll reopen while we find a proper workaround

**Comment by allanberrocal** [comment]
Hi, I'm having the issue when sending push notification to iOS. 
Error 

`node-pre-gyp ERR! parse-server-push-adapter APNS APNS error transmitting to device ...f6r4... with status 400 and reason BadDeviceToken`

We have the latest version of parse. One production certificate, no dev. 
Push notifications worked well for iOS devices for a day, but not afterwards. Not even new installations of iOS are not receiving the notifications. 

We were using the token-based authentication (w/ .p8 file) when Push stopped working. Them moved to signing certificate (w/ .p12 file), but with no better luck. Anyone is having or has had this issue? Ideas? 

Thank you

**Comment by davimacedo** [comment]
Can you please open a new issue and fill out the template?
